adapters/celestia/src/verifier/address.rs

```diff
@@ -1,56 +1,90 @@
 use std::fmt::{Display, Formatter};
 use std::str::FromStr;
 
+use bech32::WriteBase32;
 use borsh::{BorshDeserialize, BorshSerialize};
 use serde::{Deserialize, Serialize};
 use sov_rollup_interface::AddressTrait;
 use thiserror::Error;
 
+/// Human Readable Part: "celestia" for Celestia network
 const HRP: &str = "celestia";
+/// Bech32 variant is used for Celestia and CosmosSDK
+const VARIANT: bech32::Variant = bech32::Variant::Bech32;
 
+/// Representation of the address in the Celestia network
+/// https://github.com/celestiaorg/celestia-specs/blob/e59efd63a2165866584833e91e1cb8a6ed8c8203/src/specs/data_structures.md#address
+/// Spec says: "Addresses have a length of 32 bytes.", but in reality it is 32 `u5` elements, which can be compressed as 20 bytes.
 #[derive(
     Debug, PartialEq, Clone, Eq, Serialize, Deserialize, BorshDeserialize, BorshSerialize, Hash,
 )]
-// Raw ASCII bytes, including HRP
-// TODO: https://github.com/Sovereign-Labs/sovereign-sdk/issues/469
-pub struct CelestiaAddress(Vec<u8>);
+pub struct CelestiaAddress([u8; 32]);
 
 impl AsRef<[u8]> for CelestiaAddress {
     fn as_ref(&self) -> &[u8] {
         self.0.as_ref()
     }
 }
 
+/// Decodes slice of bytes into CelestiaAddress
+/// Treats it as string if it starts with HRP and the rest is valid ASCII
+/// Otherwise just checks if it contains valid `u5` elements and has the correct length.
 impl<'a> TryFrom<&'a [u8]> for CelestiaAddress {
     type Error = anyhow::Error;
 
     fn try_from(value: &'a [u8]) -> Result<Self, Self::Error> {
-        Ok(Self(value.to_vec()))
+        if value.len() >= HRP.len() && &value[..HRP.len()] == HRP.as_bytes() && value.is_ascii() {
+            // safety, because we checked that it is ASCII
+            let s = unsafe { std::str::from_utf8_unchecked(value) };
+            return CelestiaAddress::from_str(s).map_err(|e| anyhow::anyhow!("{}", e));
+        }
+        if value.len() != 32 {
+            anyhow::bail!("An address must be 32 u5 long");
+        }
+        let mut raw_address = [0u8; 32];
+        for (idx, &item) in value.iter().enumerate() {
+            bech32::u5::try_from_u8(item)
+                .map_err(|e| anyhow::anyhow!("Element at {} is not u5: {}", idx, e))?;
+            raw_address[idx] = item;
+        }
```

Comment:
I think this decoding is incorrect. An address is just a sha256 hash of the pubkey, so its bytes are not secretly `u5`s. If we get passed a slice of 32 bytes, we shouldn't do any decoding - just a copy from slice.

Reply:
Discussed over zoom